### ISSUE TITLE
Variable of type list[list[T]] instantiated as concatenation of empty sublists with non-empty sublists is type list[list[Unknown]]

### ISSUE THREAD

**Issue by thisisrandy** [microsoft/pyright#3519]
**Describe the bug**
Code describes this most succinctly:

```python
# list[list[Unknown]] - expected, though see final example
foo = [[]] + [[1]]

# list[list[int]] - expected
foo: list[list[int]] = [[]]

# list[list[int]] - expected
foo = [[1]

# list[list[int]] - expected
foo: list[list[int]] = [[1]] + [[]]

# list[list[Unknown]] - unexpected
foo: list[list[int]] = [[]] + [[1]]

# list[list[int]] - unexpected (inconsistent with first example). perhaps this is the
# expected result, but should be consistent regardless of ordering
foo = [[1]] + [[]]
```

**To Reproduce**
See above.

**Expected behavior**
See above.

**VS Code extension or command-line**
Pylance v2022.5.3

**Comment by erictraut** [comment]
Yes, this behavior is as designed, although I can understand why it might seem inconsistent.

The examples above are very different cases that involve different type inference code paths. I've reorganized them below and given them unique names so it's easier to refer to them.

```python
u1 = [[1]]
reveal_type(u1)  # list[list[int]]

u2 = [[1]] + [[]]
reveal_type(u2)  # list[list[int]]

u3 = [[]] + [[1]]
reveal_type(u3)  # list[list[Unknown]]
```

The above three expressions are assigned to variables with no declared type. That means default inference rules apply because bidirectional type inference isn't possible.

For the addition operations in `u2` and `u3`, the types of the LHS and RHS operands are evaluated — in all four cases using default inference rules. For `u2`, the two operands of inferred to be type `list[list[int]]` and `list[list[Unknown]]`. 

The `list` class declares an `__add__` method as follows:
```python
class list(MutableSequence[_T]):
    def __add__(self, __x: list[_T]) -> list[_T]: ...
```

For `u2`, the evaluated type of this call is `list[list[int]]` (based on the LHS operand). For `u3`, the evaluated type is `list[list[Unknown]]` (once again based on the LHS operand).

```python
d1: list[list[int]] = [[]]
reveal_type(d1)  # list[list[int]]

d2: list[list[int]] = [[1]] + [[]]
reveal_type(d2)  # list[list[int]]

d3: list[list[int]] = [[]] + [[1]]
reveal_type(d3)  # list[list[Unknown]]
```

In the above three cases, the expression is assigned to a variable with a declared type. That means bidirectional type inference is applied. In the case of `d1`, the value of the expression evaluates to `list[list[int]]`. Since this is the same as the declared type, no type narrowing is applied when assigning the value back to `d1`.

For `d2`, the two operands of the `+` operation are evaluated. The LHS operand evaluates to `list[list[int]]`, and the RHS operand evaluates to `list[list[Unknown]]`. The `__add__` method is then evaluated using bidirectional type inference to inform the return result. In this case, the type of the `+` operation is `list[list[int]]`. As with `d1`, no narrowing is applied when it is assigned back to `d2` because the type matches exactly.

For `d3`, the `+` operation evaluates to `list[list[Unknown]]` (same as `u3`). When this type is assigned back to `d3`, narrowing is applied because `list[list[Unknown]]` contains an "Unknown" (an implicit `Any`), and `Any` always results in narrowing on assignment.

So all of this is working as I would expect given the type inference rules and type information provided in the `list` class.

**Comment by thisisrandy** [comment]
I understand. Thank you for the clear explanation. While the outcome seems less than ideal for this particular case, I see how deeply baked in to the inference rules it is, so I'll make the note to do `d3`-type assignments in two steps in the future.
